src/pkcs11/dev_mode_key_provisioning/aws_dev_mode_key_provisioning.c

```diff
@@ -0,0 +1,1363 @@
+ * by production ready key provisioning mechanism.
+ */
+
+/* Standard includes. */
+#include <stdio.h>
+#include <string.h>
+
+/* PKCS#11 includes. */
+#include "core_pkcs11_config.h"
+#include "core_pkcs11.h"
+
+#include "core_pkcs11_test.h"
+#include "test_param_config.h"
+
+/* Client credential includes. */
+#include "aws_clientcredential_keys.h"
+#include "iot_default_root_certificates.h"
+
+/* Key provisioning include. */
+#include "aws_dev_mode_key_provisioning.h"
+
+/* Utilities include. */
+#include "core_pki_utils.h"
+
+/* mbedTLS includes. */
+#include "mbedtls/pk.h"
+#include "mbedtls/oid.h"
+
+/* Developer convenience override, for lab testing purposes, for generating
+ * a new default key pair, regardless of whether an existing key pair is present. */
+#define keyprovisioningFORCE_GENERATE_NEW_KEY_PAIR    0
+
+/* Delay before generating new key-pair, if keyprovisioningFORCE_GENERATE_NEW_KEY_PAIR
+ * is enabled. This is to avoid possible race-condition (due to device reset) between
+ * execution of an existing image on device generates key-pair on device and flashing of
+ * new image on device. */
+#ifndef keyprovisioningDELAY_BEFORE_KEY_PAIR_GENERATION_SECS
+    #define keyprovisioningDELAY_BEFORE_KEY_PAIR_GENERATION_SECS    10
+#endif
+
+#ifndef DEV_MODE_KEY_PROVISIONING_PRINT
+    #define DEV_MODE_KEY_PROVISIONING_PRINT( ... )
+#endif
+
+/* Internal structure for parsing RSA keys. */
+
+/* Length parameters for importing RSA-2048 private keys. */
+#define MODULUS_LENGTH        pkcs11RSA_2048_MODULUS_BITS / 8
+#define E_LENGTH              3
+#define D_LENGTH              pkcs11RSA_2048_MODULUS_BITS / 8
+#define PRIME_1_LENGTH        128
+#define PRIME_2_LENGTH        128
+#define EXPONENT_1_LENGTH     128
+#define EXPONENT_2_LENGTH     128
+#define COEFFICIENT_LENGTH    128
+
+/* Adding one to all of the lengths because ASN1 may pad a leading 0 byte
+ * to numbers that could be interpreted as negative */
+typedef struct RsaParams_t
+{
+    CK_BYTE modulus[ MODULUS_LENGTH + 1 ];
+    CK_BYTE e[ E_LENGTH + 1 ];
+    CK_BYTE d[ D_LENGTH + 1 ];
+    CK_BYTE prime1[ PRIME_1_LENGTH + 1 ];
+    CK_BYTE prime2[ PRIME_2_LENGTH + 1 ];
+    CK_BYTE exponent1[ EXPONENT_1_LENGTH + 1 ];
+    CK_BYTE exponent2[ EXPONENT_2_LENGTH + 1 ];
+    CK_BYTE coefficient[ COEFFICIENT_LENGTH + 1 ];
+} RsaParams_t;
+
+/* Internal structure for capturing the provisioned state of the host device. */
+typedef struct ProvisionedState_t
+{
+    CK_OBJECT_HANDLE xPrivateKey;
+    CK_OBJECT_HANDLE xClientCertificate;
+    CK_OBJECT_HANDLE xPublicKey;
+    uint8_t * pucDerPublicKey;
+    uint32_t ulDerPublicKeyLength;
+    char * pcIdentifier; /* The token label. On some devices, a unique device
+                          * ID might be stored here which can be used as a field
+                          * in the subject of the device certificate. */
+} ProvisionedState_t;
+
+/* This function can be found in libraries/3rdparty/mbedtls_utils/mbedtls_utils.c. */
```

Comment:
This path is in amazon-freertos but users can be running some other project which may or may not have this file.

Reply:
Update the comment to 
```/* This helper function can be found in mbedtls utils pem2der.c.
 * (https://github.com/ARMmbed/mbedtls/blob/development/programs/util/pem2der.c#L75) */
```
Mbedtls is required to verify and provisioning the result of PKCS11 API in pkcs11 test.